### ISSUE TITLE
Infinite loops

### ISSUE THREAD

**Issue by srey110** [mort666/RSAEuro#1]
in CentOS-7-x86_64 function R_GeneratePEMKeys() ->GeneratePrime ()->NN_Mod ()->NN_Div ()
nn.c line 318 infinite loops!

**Comment by mort666** [comment]
Could you provide an example of the code that called this to replicate? Or are you using one of the example/demos to replicate this?

**Comment by srey110** [comment]
i changed global.h  now it work fine.
example code:

```
if (usKeyLens < RSAKeyMinLen
    || usKeyLens > RSAKeyMaxLen)
{
    Q_Printf("rsa key lens %d error.key lens must in[512 - 1024]", usKeyLens);

    return Q_RTN_FAILE;
}

R_RSA_PROTO_KEY stProtoKey;

stProtoKey.bits = usKeyLens;
stProtoKey.useFermat4 = 1;

R_RandomCreate(&m_stRandom);

return R_GeneratePEMKeys(&m_stPublicKey, &m_stPrivateKey, &stProtoKey, &m_stRandom);
```

declare:
R_RANDOM_STRUCT m_stRandom;
R_RSA_PUBLIC_KEY m_stPublicKey;    /\* RSA public key _/  
R_RSA_PRIVATE_KEY m_stPrivateKey;  /_ RSA private key */

global.h:

/\* POINTER defines a generic pointer type */
typedef unsigned char *POINTER;

/\* UINT2 defines a two byte word */
typedef unsigned short UINT2;

/\* UINT4 defines a four byte word */
typedef unsigned int UINT4;

/\* BYTE defines a unsigned character */
typedef unsigned char BYTE;

/\* internal signed value */
typedef signed int signeddigit;

**Comment by mort666** [comment]
Hmm.. 

Not sure that is a good idea, you've changed a 4byte word int (32-bit int) into a 2 byte word int (16-bit int).

**Comment by srey110** [comment]
expecting your solution.

**Comment by srey110** [comment]
i get off work, bye

**Comment by mort666** [comment]
Taking a look, been a long time since I've looked at this code, trying to replicate..

**Comment by yusefmaali** [comment]
TL;DR;
on 32 bit processor the library works as expected, on 64 bit processor the library stops to work.
The problem is essentially the UINT4 and signeddigit that on 64 bit processor get an extra 4 bytes.

------

I've got the very same problem using the library into an iOS application.
This is an old B2B application compiled only against armv7 architecture (32bit).
I was called to update the application to the current iOS version, that requires the support to arm64 (64 bit) architecture.
When the application runs on recent 64-bit proc, the library got the issue described in the first comment.

Adjusting the types in the global.h file, restore the wanted type length and the library starts to work again.
